File: src/ia_dev_env/assembler/hooks_assembler.py
from __future__ import annotations

import logging
import shutil
from pathlib import Path
from typing import List

from ia_dev_env.domain.stack_mapping import get_hook_template_key
from ia_dev_env.models import ProjectConfig
from ia_dev_env.template_engine import TemplateEngine

logger = logging.getLogger(__name__)


class HooksAssembler:
    """Copies post-compile-check.sh for compiled languages."""

    def __init__(self, resources_dir: Path) -> None:
        self._resources_dir = resources_dir

    def assemble(
        self,
        config: ProjectConfig,
        output_dir: Path,
        engine: TemplateEngine,
    ) -> List[Path]:
        """Copy hook scripts for compiled languages."""
        key = get_hook_template_key(
            config.language.name,
            config.framework.build_tool,
        )
        if not key:
            logger.info("No compile hook for %s, skipping.", config.language.name)
            return []
        hook_src = self._resources_dir / "hooks-templates" / key / "post-compile-check.sh"
        if not hook_src.is_file():
            logger.warning("Hook template not found: %s", hook_src)
            return []
        return self._copy_hook(hook_src, output_dir)

    def _copy_hook(self, hook_src: Path, output_dir: Path) -> List[Path]:
        hooks_dir = output_dir / "hooks"
        hooks_dir.mkdir(parents=True, exist_ok=True)
        dest = hooks_dir / "post-compile-check.sh"
        shutil.copy2(str(hook_src), str(dest))
        dest.chmod(dest.stat().st_mode | 0o111)
        logger.info("Copied %s", dest)
        return [dest]

File: src/assembler/hooks-assembler.ts
/**
 * HooksAssembler — copies post-compile hook scripts for compiled languages.
 *
 * Migrated from Python `assembler/hooks_assembler.py` (48 lines).
 * Template key resolution lives in {@link ../domain/stack-mapping.ts}.
 *
 * @remarks
 * Hook scripts are copied verbatim — no template placeholder replacement.
 * The `engine` parameter is accepted for API uniformity but is not used.
 *
 * @module
 */
import * as fs from "node:fs";
import * as path from "node:path";
import type { ProjectConfig } from "../models.js";
import type { TemplateEngine } from "../template-engine.js";
import { getHookTemplateKey } from "../domain/stack-mapping.js";

const HOOKS_DIR = "hooks";
const HOOK_FILENAME = "post-compile-check.sh";
const HOOKS_TEMPLATES_DIR = "hooks-templates";
const EXECUTE_BITS = 0o111;

/** Assembles post-compile hook scripts for compiled languages. */
export class HooksAssembler {
  /** Copy hook scripts for compiled languages to the output directory. */
  assemble(
    config: ProjectConfig,
    outputDir: string,
    resourcesDir: string,
    _engine: TemplateEngine,
  ): string[] {
    const key = getHookTemplateKey(
      config.language.name,
      config.framework.buildTool,
    );
    if (key === "") return [];
    const hookSrc = path.join(
      resourcesDir, HOOKS_TEMPLATES_DIR, key, HOOK_FILENAME,
    );
    if (!fs.existsSync(hookSrc)) return [];
    return this.copyHook(hookSrc, outputDir);
  }

  private copyHook(hookSrc: string, outputDir: string): string[] {
    const hooksDir = path.join(outputDir, HOOKS_DIR);
    fs.mkdirSync(hooksDir, { recursive: true });
    const dest = path.join(hooksDir, HOOK_FILENAME);
    fs.copyFileSync(hookSrc, dest);
    const currentMode = fs.statSync(dest).mode;
    fs.chmodSync(dest, currentMode | EXECUTE_BITS);
    return [dest];
  }
}
